sdk/ml/azure-ai-ml/azure/ai/ml/_utils/_cache_utils.py

```diff
@@ -0,0 +1,76 @@
+# ---------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# ---------------------------------------------------------
+import uuid
+from collections import defaultdict
+from typing import Union, Tuple, List
+
+from azure.ai.ml.constants._common import AzureMLResourceType
+from azure.ai.ml.entities import Component
+from azure.ai.ml.entities._builders import BaseNode
+
+
+class CachedNodeResolver(object):
+    """Class to resolve component in nodes with cached component resolution results.
+    """
+    _ANONYMOUS_HASH_PREFIX = "anonymous_component:"
+    _YAML_SOURCE_PREFIX = "yaml_source:"
+    _ORIGIN_PREFIX = "origin:"
+
+    def __init__(self, resolver):
+        self._resolver = resolver
+        self._component_resolution_cache = {}
+        self._new_components: List[Tuple[str, Component]] = []
```

Comment:
Maybe a more meaningful name or comments? Not sure if I could get the usage by its name now.

Reply:
Good catch! Updated.